### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by dapatil211**

### CODE REVIEW

File: hive/agents/dqn.py [hunk 1]
```diff
@@ -314,9 +319,20 @@ def update(self, update_info):
...
+            self.run_debugging(observations=copy.deepcopy(current_state_inputs[0].numpy()),
+                               model=copy.deepcopy(self._qnet),
+                               labels=copy.deepcopy(q_targets),
+                               predictions=copy.deepcopy(pred_qvals.detach()),
+                               loss=copy.deepcopy(self._loss_fn),
+                               opt=copy.deepcopy(self._optimizer),
+                               actions=copy.deepcopy(actions),
+                               done = copy.deepcopy(update_info["done"])
+                               )
+
```
Comment (dapatil211) [inline hive/agents/dqn.py:333]: Is deepcopy necessary? Lazy deepcopy?

File: hive/debugger/Checkers/NN_checkers/BiasCheck.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+from hive.debugger.DebuggerInterface import DebuggerInterface
+from hive.debugger.utils.model_params_getters import get_model_weights_and_biases
+
+
+class BiasCheck(DebuggerInterface):
+
+    def __init__(self, check_period):
+        super().__init__()
+        self.check_type = "Bias"
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/BiasCheck.py]: pass to super constructor

File: hive/debugger/Checkers/NN_checkers/BiasCheck.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        self.check_period = check_period
+        self.iter_num = -1
+
+    def run(self, model):
+        error_msg = list()
+        _, initial_biases = get_model_weights_and_biases(model)
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/BiasCheck.py]: This will throw an error on gpu because you need to copy the weights to cpu.
`value.cpu().numpy()`

Comment (dapatil211) [reply hive/debugger/Checkers/NN_checkers/BiasCheck.py]: Also see if you can cache the result of this for all checks that use it. Maybe cache based on step number.

File: hive/debugger/Checkers/NN_checkers/BiasCheck.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        if not initial_biases:
+            error_msg.append(self.main_msgs['need_bias'])
+        else:
+            checks = []
+            for b_name, b_array in initial_biases.items():
+                checks.append(np.sum(b_array) == 0.0)
+            if not np.all(checks):
+                error_msg.append(self.main_msgs['zero_bias'])
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/BiasCheck.py]: Can you verify the logic for this?

Comment (dapatil211) [reply hive/debugger/Checkers/NN_checkers/BiasCheck.py]: Can you short circuit the loop? Or if you want to go through the whole loop, provide the name of the weight that is 0. eg
```
for b_name, b_array:
   if b_array == 0:
       error(b_name is bad)
```

File: hive/debugger/Checkers/NN_checkers/LossCheck.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        super().__init__()
+        self.check_type = "Loss"
+        self.check_period = check_period
+        self.iter_num = -1
+
+    def run(self, labels, predictions, loss, model):
+        error_msgs = list()
+        losses = []
+        n = self.config["init_loss"]["size_growth_rate"]
+        while n <= (self.config["init_loss"]["size_growth_rate"] * self.config["init_loss"]["size_growth_iters"]):
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/LossCheck.py]: Maybe put these as part of the constructor with default values?

File: hive/debugger/Checkers/NN_checkers/LossCheck.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+from hive.debugger.DebuggerInterface import DebuggerInterface
+import tensorflow as tf
+from hive.debugger.utils.model_params_getters import get_loss, get_model_weights_and_biases
+from hive.utils.torch_utils import numpify
+
+
+class LossCheck(DebuggerInterface):
+
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/LossCheck.py]: Add docstring for each check with short summary of what it does

File: hive/debugger/Checkers/NN_checkers/LossCheck.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        self.iter_num = -1
+
+    def run(self, labels, predictions, loss, model):
+        error_msgs = list()
+        losses = []
+        n = self.config["init_loss"]["size_growth_rate"]
+        while n <= (self.config["init_loss"]["size_growth_rate"] * self.config["init_loss"]["size_growth_iters"]):
+            derived_batch_y = np.concatenate(n * [labels], axis=0)
+            derived_predictions = np.concatenate(n * [numpify(predictions)], axis=0)
+            loss_value = float(get_loss(derived_predictions, derived_batch_y, loss))
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/LossCheck.py]: differentiate between loss and loss_fn

File: hive/debugger/Checkers/NN_checkers/ProperFittingCheck.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        opt = opt.__class__(model.parameters(),)
+        error_msg = list()
+        derived_batch_x = observations
+        derived_batch_y = labels
+
+        real_losses = []
+        model.train(True)
+        for i in range(self.config["total_iters"]):
+            print(i)
```
Comment (dapatil211) [inline hive/debugger/Checkers/NN_checkers/ProperFittingCheck.py]: remove
